In [1]:
from Bio import SeqIO
import pandas as pd

pd.set_option('display.max_columns', None)

In [8]:
fastaseq = SeqIO.parse("./aligned.fasta", "fasta") 

In [13]:
seqs = SeqIO.parse("./sequences.fasta", "fasta")

In [9]:
inf = []
seq = []
seqlen = []

for seq_record in fastaseq:
   # print(seq_record.id)
    inf.append(seq_record.id)
  #  print(repr(seq_record.seq))
    seq.append(seq_record.seq)
  #  print(len(seq_record),"\n")
    seqlen.append(len(seq_record))
    

In [15]:
inf

['Japan/SZ-NIG-Y223038/2022',
 'SYC/SS10663/2022',
 'ZAF/NHLS-UCT-GS-CE61/2022',
 'KEN/C121047/2022',
 'USA/S615AM45_538546/2023',
 'VNM/OUCRU-NHTD3617/2022',
 'ZAF/NHLS-UCT-GS-CD94/2022',
 'ZAF/NHLS-UCT-PA-J084/2022',
 'ZAF/NHLS-UCT-LA-ZB82/2022',
 'KEN/SS11083/2022',
 'USA/USA_UT-IDEL-2023021544_2023_NC_045512.2_NC_045512.2/2023',
 'KEN/C120125/2022',
 'KEN/SS11002/2022',
 'Switzerland/GE-HUG-40509532/2023',
 'ZAF/NHLS-UCT-GP-U285/2022',
 'England/QEUH-326BE15A/2023',
 'USA/VI-CDC-2-6875014/2023',
 'Switzerland/BS-UHB-USB_2023_01_30_f6430886/2023',
 'OQ155262',
 'THA/CONI-3921/2022',
 'USA/AK-USAFSAM-S21082/2023',
 'ZAF/NHLS-UCT-GP-U372/2022',
 'BHR/660304023/2022',
 'USA/UT-CDC-VSX-A079256/2023',
 'OX449278',
 'NZL/22CV5342/2022',
 'Canisfamiliaris/NAM/1256/2022',
 'NZL/22CH0368/2022',
 'USA/CA-CDC-STM-NBU4XHDMV/2022',
 'THA/CONI-3908/2022',
 'USA/NC-CDC-LC0895050/2022',
 'IRQ/BG-USAFSAM-S20950/2022',
 'OQ346311',
 'OQ346310',
 'BHR/730105709/2022',
 'ZAF/NHLS-UCT-LA-ZC13/2022',
 'C

In [16]:
inf2 = []
seq2 = []
seqlen2 = []

for seq_record in seqs:
   # print(seq_record.id)
    inf2.append(seq_record.id)
  #  print(repr(seq_record.seq))
    seq2.append(seq_record.seq)
  #  print(len(seq_record),"\n")
    seqlen2.append(len(seq_record))

In [5]:
metadata = pd.read_csv("./metadata.tsv", delimiter="\t")
metadata.head()

strain virus   gisaid_epi_isl genbank_accession  \
0   BRA/LACENAL-270228603/2022  ncov                ?          OP107768   
1           BHR/310034248/2022  ncov                ?          OP889767   
2  BGD/icddrb_TND_02_0878/2022  ncov                ?          OQ075408   
3          DJI/NAMRU3_CL1/2020  ncov  EPI_ISL_3845355          MZ970746   
4           BHR/890901686/2022  ncov                ?          OP733086   

  genbank_accession_rev sra_accession        date         region     country  \
0            OP107768.1           NaN  2022-07-01  South America      Brazil   
1            OP889767.1           NaN  2022-11-18           Asia     Bahrain   
2            OQ075408.1   SRR22799528  2022-09-25           Asia  Bangladesh   
3            MZ970746.1           NaN  2020-04-28         Africa    Djibouti   
4            OP733086.1           NaN  2022-10-16           Asia     Bahrain   

     division  ... QC_frame_shifts QC_stop_codons QC_overall_score  \
0      Brazil  ...            good           good         0.000000   
1     Bahrain  ...            good           good        29.340278   
2  Bangladesh  ...            good           good         0.000000   
3    Djibouti  ...            good           good         0.000000   
4     Bahrain  ...            good           good        34.151111   

  QC_overall_status frame_shifts  \
0              good          NaN   
1              good          NaN   
2              good          NaN   
3              good          NaN   
4          mediocre          NaN   

                                           deletions insertions  \
0  686-694,11288-11296,21633-21641,21765-21770,28...        NaN   
1    11288-11296,21633-21641,28362-28370,29734-29759        NaN   
2  11288-11296,21633-21641,21992-21994,28362-2837...        NaN   
3                                                NaN        NaN   
4                11288-11296,21633-21641,28362-28370        NaN   

                                       substitutions  \
0  C241T,T670G,C2790T,C3037T,G4184A,C4321T,C9344T...   
1  C147T,C241T,T670G,C2790T,C3037T,C3796T,C3927T,...   
2  C241T,A405G,T670G,C2790T,G4184A,C4321T,C9344T,...   
3  C241T,C1059T,C3037T,C14408T,C14576T,A23403G,G2...   
4  A437G,T670G,C3037T,T3200A,T3226A,T3488A,C3796T...   

                                     aaSubstitutions clock_deviation  
0  E:T9I,M:D3N,M:Q19E,M:A63T,N:P13L,N:R203K,N:G20...             0.0  
1  E:T9I,E:T11A,M:Q19E,M:A63T,N:P13L,N:R203K,N:G2...             7.0  
2  E:T9I,E:T11A,M:Q19E,M:A63T,N:R203K,N:G204R,N:S...             0.0  
3  ORF1a:T265I,ORF1b:P314L,ORF1b:A370V,ORF3a:Q57H...             0.0  
4  E:T9I,E:T11A,M:Q19E,M:A63T,N:P13L,N:A173V,N:R2...           -11.0  

[5 rows x 56 columns]

In [6]:
metadata.columns

Index(['strain', 'virus', 'gisaid_epi_isl', 'genbank_accession',
       'genbank_accession_rev', 'sra_accession', 'date', 'region', 'country',
       'division', 'location', 'region_exposure', 'country_exposure',
       'division_exposure', 'segment', 'length', 'host', 'age', 'sex',
       'Nextstrain_clade', 'pango_lineage', 'GISAID_clade', 'originating_lab',
       'submitting_lab', 'authors', 'url', 'title', 'paper_url',
       'date_submitted', 'date_updated', 'sampling_strategy', 'database',
       'Nextclade_pango', 'immune_escape', 'ace2_binding', 'missing_data',
       'divergence', 'nonACGTN', 'coverage', 'rare_mutations',
       'reversion_mutations', 'potential_contaminants', 'QC_missing_data',
       'QC_mixed_sites', 'QC_rare_mutations', 'QC_snp_clusters',
       'QC_frame_shifts', 'QC_stop_codons', 'QC_overall_score',
       'QC_overall_status', 'frame_shifts', 'deletions', 'insertions',
       'substitutions', 'aaSubstitutions', 'clock_deviation'],
      dtype='object')

In [7]:
len(metadata["strain"])

2925

In [8]:
len(inf)

2924

In [12]:
len(inf2)

2925

In [9]:
i=0
for strain in metadata["strain"]:
    if strain in inf:
        i+=1
i

2924

In [13]:
i=0
for strain in metadata["strain"]:
    if strain in inf2:
        i+=1
i

2925

Now we're talking.

In [14]:
metadata["strain"].value_counts()

BRA/LACENAL-270228603/2022                             1
Switzerland/TI-EOC-32_27716877/2023                    1
THA/MTM_01_573/2022                                    1
TWN/CGMH-CGU-58/2021                                   1
Switzerland/VS-ICH-2221124790/2022                     1
                                                      ..
KEN/SS11196/2022                                       1
Netherlands/Gelderland_104/2020                        1
MYS/IIUM_9992_ARTIC_medaka_1-29896/2021                1
KEN/SS11208/2022                                       1
IMS-10375-CVDP-65F86297-332E-4E2B-A87F-FC7B6BF2342C    1
Name: strain, Length: 2925, dtype: int64

In [15]:
metadata[["strain"]].describe()

strain
count                         2925
unique                        2925
top     BRA/LACENAL-270228603/2022
freq                             1

In [16]:
inf[0:10]

['Japan/SZ-NIG-Y223038/2022',
 'SYC/SS10663/2022',
 'ZAF/NHLS-UCT-GS-CE61/2022',
 'KEN/C121047/2022',
 'USA/S615AM45_538546/2023',
 'VNM/OUCRU-NHTD3617/2022',
 'ZAF/NHLS-UCT-GS-CD94/2022',
 'ZAF/NHLS-UCT-PA-J084/2022',
 'ZAF/NHLS-UCT-LA-ZB82/2022',
 'KEN/SS11083/2022']

In [17]:
inf2[0:10]

['USA/S615AM45_538546/2023',
 'KEN/C121047/2022',
 'ZAF/NHLS-UCT-GS-CE61/2022',
 'ZAF/NHLS-UCT-GS-CD94/2022',
 'VNM/OUCRU-NHTD3617/2022',
 'Japan/SZ-NIG-Y223038/2022',
 'SYC/SS10663/2022',
 'ZAF/NHLS-UCT-PA-J084/2022',
 'ZAF/NHLS-UCT-LA-ZB82/2022',
 'KEN/SS11083/2022']

In [21]:
for i in inf:
    if i not in inf2:
        print(i)

In [22]:
for i in inf2:
    if i not in inf:
        print(i)

URY/6553/2021


## Observation 1

As we can see, the *aligned.fasta* and *sequences.fatsa* are almost the same. The only strain missing in  *aligned.fasta* is **URY/6553/2021**. Besides, strains in *sequences.fatsa* are identical to the strains in the *metadata.tsv*: 2925 unique values.

In [24]:
metadata.loc[metadata["strain"]=="URY/6553/2021"]

strain virus gisaid_epi_isl genbank_accession  \
1889  URY/6553/2021  ncov              ?          OM971721   

     genbank_accession_rev sra_accession        date         region  country  \
1889            OM971721.1           NaN  2021-12-06  South America  Uruguay   

     division  ... QC_frame_shifts QC_stop_codons QC_overall_score  \
1889  Uruguay  ...             NaN            NaN              NaN   

     QC_overall_status frame_shifts  deletions insertions substitutions  \
1889               NaN          NaN        NaN        NaN           NaN   

     aaSubstitutions clock_deviation  
1889             NaN               ?  

[1 rows x 56 columns]

In [26]:
strains = pd.DataFrame(inf2, columns=["strain"])
strains

strain
0                              USA/S615AM45_538546/2023
1                                      KEN/C121047/2022
2                             ZAF/NHLS-UCT-GS-CE61/2022
3                             ZAF/NHLS-UCT-GS-CD94/2022
4                               VNM/OUCRU-NHTD3617/2022
...                                                 ...
2920  IMS-10261-CVDP-3503F74E-7B1D-47CF-BB1B-03A0ABA...
2921  IMS-10136-CVDP-9E54291A-B98D-4F0B-936D-A08FA15...
2922  IMS-10370-CVDP-2E339494-D0DE-46AC-B059-159C0EF...
2923  IMS-10146-CVDP-F89F97C1-C1CD-48F1-98D2-869E787...
2924  IMS-10087-CVDP-093C7DFC-4022-41A1-B274-21FFF6F...

[2925 rows x 1 columns]

In [27]:
strains.describe()

strain
count                       2925
unique                      2925
top     USA/S615AM45_538546/2023
freq                           1

## EDA

In [28]:
metadata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2925 entries, 0 to 2924
Data columns (total 56 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   strain                  2925 non-null   object 
 1   virus                   2925 non-null   object 
 2   gisaid_epi_isl          2925 non-null   object 
 3   genbank_accession       2925 non-null   object 
 4   genbank_accession_rev   2802 non-null   object 
 5   sra_accession           1346 non-null   object 
 6   date                    2925 non-null   object 
 7   region                  2925 non-null   object 
 8   country                 2925 non-null   object 
 9   division                2925 non-null   object 
 10  location                379 non-null    object 
 11  region_exposure         2925 non-null   object 
 12  country_exposure        2925 non-null   object 
 13  division_exposure       2925 non-null   object 
 14  segment                 2925 non-null   

In [30]:
metadata

strain virus  \
0                            BRA/LACENAL-270228603/2022  ncov   
1                                    BHR/310034248/2022  ncov   
2                           BGD/icddrb_TND_02_0878/2022  ncov   
3                                   DJI/NAMRU3_CL1/2020  ncov   
4                                    BHR/890901686/2022  ncov   
...                                                 ...   ...   
2920  IMS-10582-CVDP-970C8B96-B99D-4610-B497-FEAF6D3...  ncov   
2921  IMS-10375-CVDP-EF1A8D0C-2B58-470C-99E2-5B5C29D...  ncov   
2922  IMS-10722-CVDP-52C559C3-7225-4089-BB4E-4CF1B3C...  ncov   
2923  IMS-10594-CVDP-F1DB0427-D718-4517-90C2-C141C9C...  ncov   
2924  IMS-10375-CVDP-65F86297-332E-4E2B-A87F-FC7B6BF...  ncov   

       gisaid_epi_isl genbank_accession genbank_accession_rev sra_accession  \
0                   ?          OP107768            OP107768.1           NaN   
1                   ?          OP889767            OP889767.1           NaN   
2                   ?          OQ075408            OQ075408.1   SRR22799528   
3     EPI_ISL_3845355          MZ970746            MZ970746.1           NaN   
4                   ?          OP733086            OP733086.1           NaN   
...               ...               ...                   ...           ...   
2920                ?                 ?                   NaN             ?   
2921                ?                 ?                   NaN             ?   
2922                ?                 ?                   NaN             ?   
2923                ?                 ?                   NaN             ?   
2924                ?                 ?                   NaN             ?   

            date         region     country    division location  \
0     2022-07-01  South America      Brazil      Brazil      NaN   
1     2022-11-18           Asia     Bahrain     Bahrain      NaN   
2     2022-09-25           Asia  Bangladesh  Bangladesh      NaN   
3     2020-04-28         Africa    Djibouti    Djibouti      NaN   
4     2022-10-16           Asia     Bahrain     Bahrain      NaN   
...          ...            ...         ...         ...      ...   
2920  2023-02-28         Europe     Germany           ?        ?   
2921  2022-09-29         Europe     Germany           ?        ?   
2922  2022-10-16         Europe     Germany           ?        ?   
2923  2022-11-03         Europe     Germany           ?        ?   
2924  2022-09-26         Europe     Germany           ?        ?   

     region_exposure country_exposure division_exposure segment  length  \
0      South America           Brazil            Brazil  genome   29776   
1               Asia          Bahrain           Bahrain  genome   29724   
2               Asia       Bangladesh        Bangladesh  genome   29721   
3             Africa         Djibouti          Djibouti  genome   29782   
4               Asia          Bahrain           Bahrain  genome   29466   
...              ...              ...               ...     ...     ...   
2920          Europe          Germany                 ?  genome   29847   
2921          Europe          Germany                 ?  genome   29850   
2922          Europe          Germany                 ?  genome   29724   
2923          Europe          Germany                 ?  genome   29694   
2924          Europe          Germany                 ?  genome   29848   

              host age   sex Nextstrain_clade pango_lineage GISAID_clade  \
0     Homo sapiens   ?     ?    22B (Omicron)        BA.5.1            ?   
1     Homo sapiens   ?     ?    22D (Omicron)          BN.1            ?   
2     Homo sapiens   ?  Male    22F (Omicron)         XBB.1            ?   
3     Homo sapiens   ?     ?              20C           B.1            ?   
4     Homo sapiens   ?     ?    22D (Omicron)          BN.1            ?   
...            ...  ..   ...              ...           ...          ...   
2920             ?   ?     ?    23A (Omicron)       XBB.1.5          

In [31]:
metadata["virus"].value_counts()

ncov    2925
Name: virus, dtype: int64

In [32]:
seq[0:5]

[Seq('------------------------TAACAAACCAACCAACTTTCGATCTCTTGT...---'),
 Seq('------------------------------------------------------...---'),
 Seq('------------------------------------------------------...---'),
 Seq('--------------------------------------ACTTTCGATCTCTTGT...---'),
 Seq('------------------------------------------------------...---')]

In [33]:
seq2[0:5]

[Seq('AGATCTGTTCTCTAAACGAACTTTAAAATCTGTGTGGCTGTCACTCGGCTGCAT...TCA'),
 Seq('ACTTTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAATCTGTGTGGC...AAT'),
 Seq('GTTCGCGACGTGCTCGTACGTGGCTTTGGAGACTCCGTGGAGGAGGTCTTATCA...AGT'),
 Seq('TTGTAGATCTGTTCTCTAAACGAACTTTAAAATCTGTGTGGCTGTCACTCGGCT...AGT'),
 Seq('CTTTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAATCTGTGTGGCT...AGC')]

In [34]:
seqlen[0:10]

[29903, 29903, 29903, 29903, 29903, 29903, 29903, 29903, 29903, 29903]

In [35]:
seqlen2[0:10]

[29585, 29757, 29433, 29480, 29733, 29764, 29456, 29435, 29440, 29727]

## Observation 2

So that's the crucial difference between *aligned.fasta* and *sequences.fasta*. 

In [17]:
spike_boundaries = (21563, 25384)

In [18]:
spike_proteins = []
for seq in seq2:
    spike_protein = seq[spike_boundaries[0] : spike_boundaries[1]+1]
    spike_proteins.append(spike_protein)
spike_proteins    

[Seq('CTCAATCATACACTAATTCTTTCACACGTGGTGTTTATTACCCTGACAAAGTTT...TGC'),
 Seq('AATCTTATAACCAGAACTCAATCATACACTAATTCTTTCACACGTGGTGTTTAT...AAG'),
 Seq('TGTTAATAACGCTACTAATGTTGTTATTAAAGTCTGTGAATTTCAATTTTGTAA...AGT'),
 Seq('AGAACTCAATCATACACTAATTCTTTCACACGTGGTGTTTATTACCCTGACAAA...TCA'),
 Seq('ATCTTATAACCAGAACTCAATCATACACTAATTCTTTCACACGTGGTGTTTATT...TTT'),
 Seq('CTAGTCAGTGTGTTAATCTTATAACCAGAACTCAATCATACACTAATTCTTTCA...AAC'),
 Seq('AACGCTACTAATGTTGTTATTAAAGTCTGTGAATTTCAATTTTGTAATGATCCA...TAA'),
 Seq('TAAAGTCTGTGAATTTCAATTTTGTAATGATCCATTTTTGGATGTTTATTACCA...ATG'),
 Seq('GTTATTAAAGTCTGTGAATTTCAATTTTGTAATGATCCATTTTTGGATGTTTAT...TAA'),
 Seq('ACTAGAACTCAATCATACACTAATTCTTTCACACGTGGTGTTTATTANNNNNNN...AAA'),
 Seq('AGAACTCAATCATACACTAATTCTTTCACACGTGGTGTTTATTANNNNNNNNNN...AGG'),
 Seq('AGAACTCAATCATACACTAATTCTTTCACACGTGGTGTTTATTACCCTGACAAA...CTA'),
 Seq('ACTCAATCATACACTAATTCTTTCACACGTGGTGTTTATTACCCTGACAAAGTT...CTA'),
 Seq('TTGTTTTATTGCCACTAGTCTCTAGTCAGTGTGTTAATCTTATAACCAGAACTC...GAG'),
 Seq('TTCTTGTTTTATTG

In [47]:
spike_proteins[0]

Seq('CTCAATCATACACTAATTCTTTCACACGTGGTGTTTATTACCCTGACAAAGTTT...TGC')

In [49]:
seq2[0][21563:25385]

Seq('CTCAATCATACACTAATTCTTTCACACGTGGTGTTTATTACCCTGACAAAGTTT...TGC')

In [19]:
seq2[0][21563:25385]==spike_proteins[0]

True

It worked. Now we have to allign them. But first, save them as *fasta*: 

In [23]:
from Bio.SeqRecord import SeqRecord


seq_records = []
for i in range(len(inf)):
    record = SeqRecord(spike_proteins[i], id=inf[i])
    seq_records.append(record)
seq_records

[SeqRecord(seq=Seq('CTCAATCATACACTAATTCTTTCACACGTGGTGTTTATTACCCTGACAAAGTTT...TGC'), id='Japan/SZ-NIG-Y223038/2022', name='<unknown name>', description='<unknown description>', dbxrefs=[]),
 SeqRecord(seq=Seq('AATCTTATAACCAGAACTCAATCATACACTAATTCTTTCACACGTGGTGTTTAT...AAG'), id='SYC/SS10663/2022', name='<unknown name>', description='<unknown description>', dbxrefs=[]),
 SeqRecord(seq=Seq('TGTTAATAACGCTACTAATGTTGTTATTAAAGTCTGTGAATTTCAATTTTGTAA...AGT'), id='ZAF/NHLS-UCT-GS-CE61/2022', name='<unknown name>', description='<unknown description>', dbxrefs=[]),
 SeqRecord(seq=Seq('AGAACTCAATCATACACTAATTCTTTCACACGTGGTGTTTATTACCCTGACAAA...TCA'), id='KEN/C121047/2022', name='<unknown name>', description='<unknown description>', dbxrefs=[]),
 SeqRecord(seq=Seq('ATCTTATAACCAGAACTCAATCATACACTAATTCTTTCACACGTGGTGTTTATT...TTT'), id='USA/S615AM45_538546/2023', name='<unknown name>', description='<unknown description>', dbxrefs=[]),
 SeqRecord(seq=Seq('CTAGTCAGTGTGTTAATCTTATAACCAGAACTCAATCATACACTAATTCTTTCA

In [128]:
len(seq_records)

2924

In [32]:
SeqIO.write(seq_records, "spike_proteins.fasta", "fasta")

2924

In [33]:
#Test
test = SeqIO.parse("./spike_proteins.fasta", "fasta") 
test

## MSA

In [7]:
from Bio import AlignIO
msa = AlignIO.read("./spike_proteins.fasta", "fasta")

ValueError: Sequences must all be the same length

# Notiz für Jan:
Aus irgendeinem Grund kann ich die gespeicherten spikes als fasta-Datei nicht alignen. Ich verstehe noch nicht, was da schief läuft.

## Indels Check 

In [13]:
len(str(seq[0])) #length of a string (all the same)

29903

In [19]:
29903*0.2  #how much is 20%

5980.6

In [14]:
str(seq[0]).count("-") #example

139

In [23]:
many_indels = []
for s in seq:
    number_idels = str(s).count("-")
    if number_idels >= 5980:
        many_indels.append(s)
len(many_indels)

0

Well, in this case problem solved.